### PR TITLE
(DOCSP-26115): Flutter copy data to a new realm

### PR DESCRIPTION
**Pull request by mongodben**
## Pull Request Info

Flutter copy data to a new realm

### Jira

- https://jira.mongodb.org/browse/DOCSP-26115

### Staged Changes

- [Configure and Open a Realm (Flutter)](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-26115/sdk/flutter/realm-database/configure-and-open/)

### Reminder Checklist

If your PR modifies the docs, you might need to also update some corresponding
pages. Check if completed or N/A.

- [x] Create Jira ticket for corresponding docs-app-services update(s), if any
- [x] Checked/updated Admin API
- [x] Checked/updated CLI reference

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/flutter/realm-files/configure-and-open.txt [hunk 1]
```diff
@@ -187,3 +187,57 @@ Close a Realm
...
          Realm.shutdown();
+
+.. _flutter-copy-data-into-new-realm:
+
+Copy Data into a New Realm
+--------------------------
+
+To copy data from an existing realm to a new realm with different
+configuration options, pass the new configuration to
+:flutter-sdk:`Realm.writeCopy() <TODO: add link>`.
```
Comment (desistefanova) [inline source/sdk/flutter/realm-files/configure-and-open.txt:198]:
```suggestion
:flutter-sdk:`Realm.writeCopy() <realm/Realm/writeCopy.html>`.
```

Comment (desistefanova) [reply source/sdk/flutter/realm-files/configure-and-open.txt:198]: It is already available at https://pub.dev/documentation/realm/latest/realm/Realm/writeCopy.html

File: source/sdk/flutter/realm-files/configure-and-open.txt [hunk 2]
```diff
@@ -187,3 +187,57 @@ Close a Realm
...
+
+In the new realm's configuration, you *must* specify the ``path``.
+
+If you write the copied realm to a realm file that already exists,
+the data is written object by object. The copy operation replaces objects
+if there already exists objects for given primary keys.
+The schemas of the realm you copy and the realm you are writing to must be
+compatible for the copy operation to succeed.
+Only objects in the schemas of both configurations are copied over.
```
Comment (desistefanova) [inline source/sdk/flutter/realm-files/configure-and-open.txt:207]: We don't support merging to existing realm. According to the API doc "The destination file should not already exist."

File: source/sdk/flutter/realm-files/configure-and-open.txt [hunk 3]
```diff
@@ -187,3 +187,57 @@ Close a Realm
...
+
+Using ``Realm.writeCopy()``, you can convert between the following
+:flutter-sdk:`Configuration <realm/Configuration-class.html>`
+types:
+
+- ``LocalConfiguration`` to ``LocalConfiguration``
+- ``FlexibleSyncConfiguration`` to ``FlexibleSyncConfiguration``
+- ``FlexibleSyncConfiguration`` to ``LocalConfiguration``
+- ``LocalConfiguration`` to read-only ``LocalConfiguration``
+- ``InMemoryConfiguration`` to ``LocalConfiguration``
```
Comment (desistefanova) [inline source/sdk/flutter/realm-files/configure-and-open.txt:217]:
```suggestion
- ``LocalConfiguration`` to read-only ``LocalConfiguration`` and vice versa
- ``InMemoryConfiguration`` to ``LocalConfiguration`` and vice versa
```

File: source/sdk/flutter/realm-files/configure-and-open.txt [hunk 4]
```diff
@@ -187,3 +187,57 @@ Close a Realm
...
+
+You **cannot** convert from a ``LocalConfiguration`` or a ``InMemoryConfiguration``
+to a ``FlexibleSyncConfiguration``.
+
+Some additional considerations to keep in mind while using ``Realm.writeTo()``:
```
Comment (desistefanova) [inline source/sdk/flutter/realm-files/configure-and-open.txt:222]:
```suggestion
Some additional considerations to keep in mind while using ``Realm.writeCopy()``:
```

File: source/sdk/flutter/realm-files/configure-and-open.txt [hunk 5]
```diff
@@ -187,3 +187,57 @@ Close a Realm
...
+#. When using synced Realm, it is required that all local changes are synchronized
+   with the server before the copy can be written.
+   This is to be sure that the file can be used as a starting point
+   for a newly installed application.
+   The method throws if there are pending uploads.
+
+The following example copies the data from a realm with a ``LocalConfiguration``
+to a new realm with a ``FlexibleSyncConfiguration``.
+
+TODO: code example
```
Comment (desistefanova) [inline source/sdk/flutter/realm-files/configure-and-open.txt:235]: This is not supported. Probably we have to remove these lines.

File: source/sdk/flutter/realm-files/configure-and-open.txt [hunk 6]
```diff
@@ -187,3 +187,57 @@ Close a Realm
...
+
+You can also :ref:`include a new encryption key <flutter-encrypt>`
+in the copied realm's configuration or remove the encryption key from the new configuration.
+
+The following example copies data from an unencrypted realm with a ``LocalConfiguration``
+to an encrypted realm with a ``FlexibleSyncConfiguration``.
+
+TODO: code example
```
Comment (desistefanova) [inline source/sdk/flutter/realm-files/configure-and-open.txt:243]: Local to Flexible is not supported. Probably you have to give another example about the encryption.

File: source/examples/generated/flutter/write_copy_test.snippet.unencrypted-to-encrypted.dart [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+// the `encryptedConfig`. The data is encrypted as part of the copying.
+unencryptedRealm.writeCopy(encryptedConfig);
+// Close the realm you just copied when you're done working with it.
+unencryptedRealm.close();
+
+// Open the new encrypted realm with `encryptedConfig`.
+final encryptedRealm = Realm(encryptedConfig);
+
+// Person object for "Harper" is in `localRealm` because
+// the data was copied over with `inMemoryRealm.writeCopy()`.
```
Comment (desistefanova) [inline source/examples/generated/flutter/write_copy_test.snippet.unencrypted-to-encrypted.dart:24]:
```suggestion
// the data was copied over with `unencryptedRealm.writeCopy()`.
```
```suggestion
// the data was copied over with `inMemoryRealm.writeCopy()`.
```

File: examples/dart/test/write_copy_test.dart [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    // the `encryptedConfig`. The data is encrypted as part of the copying.
+    unencryptedRealm.writeCopy(encryptedConfig);
+    // Close the realm you just copied when you're done working with it.
+    unencryptedRealm.close();
+
+    // Open the new encrypted realm with `encryptedConfig`.
+    final encryptedRealm = Realm(encryptedConfig);
+
+    // Person object for "Harper" is in `localRealm` because
+    // the data was copied over with `inMemoryRealm.writeCopy()`.
```
Comment (mongodben) [inline examples/dart/test/write_copy_test.dart:70]:
```suggestion
    // the data was copied over with `unencryptedRealm.writeCopy()`.
```

**Review by dacharyc — APPROVED** [review]
I like this version of this content a lot better than what I've got going on in Swift, I think. The way you've listed out the options here seems very easy to digest. I'm going to have to borrow from this to restructure the content over there.

Just a couple of grammar fixes we should make here, and one comment struck me as contradicting something higher up on the page, so we might want to reword one of those instances.

File: examples/dart/test/write_copy_test.dart [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  @PrimaryKey()
+  late String name;
+}
+
+main() {
+  test("Convert in-memory realm to local realm", () {
+    // :snippet-start: in-memory-to-local
+    // Create in-memory realm and add data to it.
+    // Note that even though the realm is in-memory, it still has a file path.
+    // This is because in-memory realms still use memory-mapped files
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/dart/test/write_copy_test.dart:18]: thx for the spot! rephrased above

**Pull request merged**